src/script/api/script_text.cpp

```diff
@@ -56,37 +57,30 @@ ScriptText::ScriptText(HSQUIRRELVM vm) :
 ScriptText::~ScriptText()
 {
 	for (int i = 0; i < SCRIPT_TEXT_MAX_PARAMETERS; i++) {
-		free(this->params[i]);
-		if (this->paramt[i] != nullptr) this->paramt[i]->Release();
+		if (std::holds_alternative<ScriptText *>(this->param[i])) std::get<ScriptText *>(this->param[i])->Release();
```

Comment:
Maybe not for this PR, but... introduce some `ScriptTextRef` class that arranges all the reference management, including move semantics? Then the destructor can go, and the reference management in `_SetDParam` can be removed too.

Reply:
Oh good idea.

Reply:
I was thinking about something along the lines of https://gist.github.com/rubidium42/47ab5af7edc1884196cfc8b3c0a3ce57, but I have not tested that beyond compile/regression testing.

Reply:
ScriptText is not tested by regression, so only actual testing could spot the missing `nullptr` check in destructor. Anyway after adding the check it seems to work fine.
For my tests I added some `Debug()` to track stuff.
With
```
local text = GSText(GSText.STR_TEST);
local param0 = GSText(GSText.STR_NUM, 10);
local param1 = GSText(GSText.STR_FOO, "foo");
text.AddParam(param1);
text.AddParam(param0);
text.SetParam(1, param1);
text.SetParam(2, param0);
```
output is (blank lines added to easily match the squirrel line)
```
dbg: [misc] constructing with 0x1800beed2e0
dbg: [misc] move constructing with 0x1800beed2e0
dbg: [misc] deleting with 0x0

dbg: [misc] constructing with 0x1800bef0f40
dbg: [misc] move constructing with 0x1800bef0f40
dbg: [misc] deleting with 0x0

dbg: [misc] constructing with 0x1800beed2e0
dbg: [misc] move assignment from 0x1800beed2e0 to 0x1800beed2e0
dbg: [misc] deleting with 0x1800beed2e0

dbg: [misc] constructing with 0x1800bef0f40
dbg: [misc] move assignment from 0x1800bef0f40 to 0x1800bef0f40
dbg: [misc] deleting with 0x1800bef0f40

dbg: [misc] deleting with 0x1800bef0f40
dbg: [misc] deleting with 0x1800beed2e0
dbg: [script] The script died unexpectedly.
```

And with
```
local text = GSText(GSText.STR_TEST);
local param0 = GSText(GSText.STR_NUM, 10);
local param1 = GSText(GSText.STR_FOO, "foo");
text.AddParam(param1);
text.AddParam(param0);
text.SetParam(2, param0);
text.SetParam(1, param1);
```
output is (blank lines added to easily match the squirrel line)
```
dbg: [misc] constructing with 0x1800c1d4770
dbg: [misc] move constructing with 0x1800c1d4770
dbg: [misc] deleting with 0x0

dbg: [misc] constructing with 0x1800c1cdff0
dbg: [misc] move constructing with 0x1800c1cdff0
dbg: [misc] deleting with 0x0

dbg: [misc] constructing with 0x1800c1d4770
dbg: [misc] move assignment from 0x1800c1cdff0 to 0x1800c1d4770
dbg: [misc] deleting with 0x1800c1cdff0

dbg: [misc] constructing with 0x1800c1cdff0
dbg: [misc] move assignment from 0x1800c1d4770 to 0x1800c1cdff0
dbg: [misc] deleting with 0x1800c1d4770

dbg: [misc] deleting with 0x1800c1d4770
dbg: [misc] deleting with 0x1800c1cdff0
dbg: [script] The script died unexpectedly.
```

Reply:
Should we upstream/merge that idea before this PR? Or shall we just put it into this PR?

Reply:
Without the merge of the 3 param arrays (and replacing them with std::variant), ScriptObjectRef doesn't seem useful as there will still be stredup/free management for char*.
Putting it in this PR just after the arrays merge seems the best solution.